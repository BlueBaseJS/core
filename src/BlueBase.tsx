import { Analytics, Logger } from './api';
import {
	ComponentCollection,
	ComponentRegistry,
	ConfigCollection,
	ConfigRegistry,
	HookNestedCollection,
	HookRegistry,
	PluginCollection,
	PluginRegistry,
	ThemeCollection,
	ThemeRegistry,
} from './registries';
import { BlueBaseProvider } from './Context';
import React from 'react';
import { ThemeProvider } from './themes';
import { getComponent } from './getComponent';
import systemHooks from './hooks';

export interface BootOptions {
	components: ComponentCollection,
	configs: ConfigCollection,
	hooks: HookNestedCollection,
	plugins: PluginCollection,
	themes: ThemeCollection,
	// routes: Plugin[]
}

export class BlueBase {

	// APIs
	public Analytics = new Analytics(this);
	public Logger = new Logger(this);

	// Registries
	public Components = new ComponentRegistry(this);
	public Configs = new ConfigRegistry(this);
	public Hooks = new HookRegistry(this);
	public Plugins = new PluginRegistry(this);
	public Themes = new ThemeRegistry(this);

	// Flags
	public booted = false;

	private bootOptions: BootOptions = {
		components: {},
		configs: {},
		hooks: {},
		plugins: [],
		themes: [],
	};

	public async boot(options?: Partial<BootOptions> & { children?: React.ReactNode }) {

		this.bootOptions = { ...this.bootOptions, ...options };

		await this.Hooks.registerNestedCollection(systemHooks);
		await this.Hooks.registerNestedCollection(this.bootOptions.hooks);

		await this.Hooks.run('bluebase.boot', this.bootOptions);

		const SystemApp = getComponent('SystemApp');

		// Set View
		// const SystemApp = this.Components.resolve('SystemApp');
		// SystemApp = await this.Hooks.run('bluebase.system.app', SystemApp);

		const BlueBaseRoot = () => (
			<BlueBaseProvider value={this}>
				<ThemeProvider>
					<SystemApp>
						{options && options.children}
					</SystemApp>
				</ThemeProvider>
			</BlueBaseProvider>
		);

		this.booted = true;

		return BlueBaseRoot;
	}
}