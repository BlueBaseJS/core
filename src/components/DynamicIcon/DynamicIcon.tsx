import { BlueBase } from '../../BlueBase';
import { BlueBaseConsumer } from '../../Context';
import React from 'react';

export interface DynamicIconProps {

	/**
	 * If value is:
	 *
	 * - component: Icon is a custom component, and looks for 'component' prop
	 * - name: Icon is an instance of BB.Components.Icon and looks for 'name' prop
	 * - image: Icon is an instance of BB.Components.Image and looks for 'source' prop
	 */
	type: 'component' | 'name' | 'image';

	/**
	 * Used when type is 'component'.
	 * Either a component or a component name (string).
	 * In case of string, it will be fetched from Component Registry.
	 */
	component?: React.ComponentType<any> | string;

	/**
	 * Used when type is 'name'.
	 * The name prop of the BB.Components.Icon component
	 */
	name?: string;

	/**
	 * Used when type is 'image'.
	 * The image source
	 */
	source?: string;

	/**
	 * Icon size. Defaults to 100
	 */
	size?: number;

	[key: string]: any;
}

/**
 * An enhanced Icon that can render any of the following:
 * - BB.Components.Icon
 * - BB.Components.Image
 * - A custom component
 */
const DynamicIcon: React.ComponentType<DynamicIconProps> = (props) => {
	const { type, component: Component, name, source, ...rest } = props;

	return (
		<BlueBaseConsumer>
			{(BB: BlueBase) => {
				let component: React.ComponentType<any>;

				if (type === 'component' && Component) {
					component = (typeof Component === 'string')
						? BB.Components.resolve(Component)
						: component = Component;

				} else if (type === 'name' && name) {
					component = BB.Components.resolve('Icon');
					rest.name = name;

				} else if (type === 'image' && source) {
					component = BB.Components.resolve('Image');
					rest.source = source;

					if (!rest.style) {
						rest.style = {};
					}

					if (rest.size) {
						rest.style.width = rest.size;
						rest.style.height = rest.size;
					}
				} else {
					return null;
				}

				return React.createElement(component, rest);
			}}
		</BlueBaseConsumer>
	);
};

DynamicIcon.defaultProps = { size: 100 };

export {
	DynamicIcon
};
