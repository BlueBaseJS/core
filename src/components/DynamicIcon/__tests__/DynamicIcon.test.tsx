import { Text, View } from 'react-native';

import { BlueBase } from '../../../BlueBase';
import { BlueBaseApp } from '../../../';
import { DynamicIcon } from '../DynamicIcon';
import React from 'react';
import { mount } from 'enzyme';
import { waitForElement } from 'enzyme-async-helpers';

describe('DynamicIcon', () => {
	test(`should return null if there is no type prop`, async () => {
		const DIcon = DynamicIcon as any;

		const wrapper = mount(
			<BlueBaseApp>
				<DIcon />
			</BlueBaseApp>
		);

		await waitForElement(wrapper, DIcon);

		expect(wrapper.find(DIcon).text()).toBe('');
	});

	test(`should render an image with default size of 100x100`, async () => {
		const wrapper = mount(
			<BlueBaseApp>
				<DynamicIcon type="image" source={{ uri: 'https://picsum.photos/200' }} />
			</BlueBaseApp>
		);

		await waitForElement(wrapper, DynamicIcon);
		const node = wrapper.find('Image').first();

		expect(node.prop('size')).toBe(100);
		expect(node.prop('style')).toMatchObject({ height: 100, width: 100 });
	});

	test(`should render an image with default size of 250x250`, async () => {
		const wrapper = mount(
			<BlueBaseApp>
				<DynamicIcon type="image" size={250} source={{ uri: 'https://picsum.photos/200' }} />
			</BlueBaseApp>
		);

		await waitForElement(wrapper, DynamicIcon);
		const node = wrapper.find('Image').first();

		expect(node.prop('size')).toBe(250);
		expect(node.prop('style')).toMatchObject({ height: 250, width: 250 });
	});

	test(`should render an image with style prop overwritten`, async () => {
		const wrapper = mount(
			<BlueBaseApp>
				<DynamicIcon
					type="image"
					style={{ height: 250, width: 250 }}
					source={{ uri: 'https://picsum.photos/200' }}
				/>
			</BlueBaseApp>
		);

		await waitForElement(wrapper, DynamicIcon);
		const node = wrapper.find('Image').first();

		expect(node.prop('size')).toBe(100);
		expect(node.prop('style')).toMatchObject({ height: 100, width: 100 });
	});

	test(`should render an image with style prop overwritten`, async () => {
		const wrapper = mount(
			<BlueBaseApp>
				<DynamicIcon
					type="image"
					style={{ height: 250, width: 250 }}
					source={{ uri: 'https://picsum.photos/200' }}
				/>
			</BlueBaseApp>
		);

		await waitForElement(wrapper, DynamicIcon);
		const node = wrapper.find('Image').first();

		expect(node.prop('size')).toBe(100);
		expect(node.prop('style')).toMatchObject({ height: 100, width: 100 });
	});

	test(`should render an custom component`, async () => {
		const CustomComponent = ({ size }: { size: number }) => (
			<View testID="custom" style={{ height: size, width: size, backgroundColor: 'red' }} />
		);

		const wrapper = mount(
			<BlueBaseApp>
				<DynamicIcon type="component" component={CustomComponent} />
			</BlueBaseApp>
		);

		await waitForElement(wrapper, DynamicIcon);
		const node = wrapper.find('[testID="custom"]').last();

		expect(node.prop('style')).toMatchObject({ height: 100, width: 100, backgroundColor: 'red' });
	});

	test(`should render an custom component that is registered in component registry`, async () => {
		const CustomComponent = ({ size }: { size: number }) => (
			<View testID="custom" style={{ height: size, width: size, backgroundColor: 'green' }} />
		);

		const BB = new BlueBase();

		const wrapper = mount(
			<BlueBaseApp BB={BB} components={{ CustomComponent }}>
				<DynamicIcon type="component" component="CustomComponent" />
			</BlueBaseApp>
		);

		await waitForElement(wrapper, DynamicIcon);
		const node = wrapper.find('[testID="custom"]').last();

		expect(node.prop('style')).toMatchObject({ height: 100, width: 100, backgroundColor: 'green' });
	});

	test(`should render an Icon component`, async () => {
		const Icon = ({ size, name }: { size: number; name: string }) => (
			<View testID="custom" style={{ height: size, width: size, backgroundColor: 'blue' }}>
				<Text>{name}</Text>
			</View>
		);

		const BB = new BlueBase();

		const wrapper = mount(
			<BlueBaseApp BB={BB} components={{ Icon }}>
				<DynamicIcon type="icon" name="bus" />
			</BlueBaseApp>
		);

		await waitForElement(wrapper, DynamicIcon);
		const node = wrapper.find('[testID="custom"]').last();

		expect(node.prop('style')).toMatchObject({ height: 100, width: 100, backgroundColor: 'blue' });
	});
});
