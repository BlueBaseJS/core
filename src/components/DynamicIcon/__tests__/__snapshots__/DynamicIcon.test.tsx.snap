// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DynamicIcon should render an Icon component 1`] = `
<View
  styles={Object {}}
>
  <View
    style={
      Object {
        "backgroundColor": "blue",
        "height": 100,
        "width": 100,
      }
    }
  >
    <Text>
      bus
    </Text>
  </View>
</View>
`;

exports[`DynamicIcon should render an custom component 1`] = `
<View
  styles={Object {}}
>
  <View
    style={
      Object {
        "backgroundColor": "red",
        "height": 100,
        "width": 100,
      }
    }
  />
</View>
`;

exports[`DynamicIcon should render an custom component that is registered in component registry 1`] = `
<View
  styles={Object {}}
>
  <View
    style={
      Object {
        "backgroundColor": "green",
        "height": 100,
        "width": 100,
      }
    }
  />
</View>
`;

exports[`DynamicIcon should render an image with default size of 100x100 1`] = `
<View
  styles={Object {}}
>
  <Image
    size={100}
    source={
      Object {
        "uri": "https://picsum.photos/200",
      }
    }
    style={
      Object {
        "height": 100,
        "width": 100,
      }
    }
    styles={Object {}}
  />
</View>
`;

exports[`DynamicIcon should render an image with default size of 250x250 1`] = `
<View
  styles={Object {}}
>
  <Image
    size={250}
    source={
      Object {
        "uri": "https://picsum.photos/200",
      }
    }
    style={
      Object {
        "height": 250,
        "width": 250,
      }
    }
    styles={Object {}}
  />
</View>
`;

exports[`DynamicIcon should render an image with style prop overwritten 1`] = `
<View
  styles={Object {}}
>
  <Image
    size={100}
    source={
      Object {
        "uri": "https://picsum.photos/200",
      }
    }
    style={
      Object {
        "height": 100,
        "width": 100,
      }
    }
    styles={Object {}}
  />
</View>
`;

exports[`DynamicIcon should render an image with style prop overwritten 2`] = `
<View
  styles={Object {}}
>
  <Image
    size={100}
    source={
      Object {
        "uri": "https://picsum.photos/200",
      }
    }
    style={
      Object {
        "height": 100,
        "width": 100,
      }
    }
    styles={Object {}}
  />
</View>
`;
